Added the Event System!

ridhwandll opened the pull request:
- The `!roll` command is now fixed. The Bot doesn't crash if values greater or less than the 32bit limits!
- Added an event system!

ridhwandll committed:
Fixed the crash of "!roll" command.


Non-timed events added!

Timed events will be added later!

ridhwandll committed:
Added event system.

Non-timed events are deleted. Only timed events! Admin can use the "!endevent" command to manually end an event. 
Removed "qDebug()" from CurrencyModule.cpp 


Fixed an issue regarding events(LowRiskLowReward)

ridhwandll committed:
Added more juice to the bot's response regarding events.

ridhwandll committed:
Added separate module for the events!

ridhwandll committed:
Added automatic restore of the changes done by events if bot crashes

Removed the "!restore" command. It is done automatically when the bot restarts! Fixed the "!help" command for the events.

FancyKillerPanda commented on res/commands.json:
```diff
@@ -270,6 +270,32 @@
+	"EVENT": {
+		"brief": "Retrieve information about the running event(s).",
+		"usage": "event",
+		"additional": "**Example: ** `!event` will retrieve information about the running event(s)",
```
This is kinda unnecessary, it is already described by the `brief` and `usage`.

ridhwandll replied:
Well...then what should the command `!help event` should display?

FancyKillerPanda replied:
Sorry I meant the `additional` bit was unnecessary.

ridhwandll replied:
Okay I will remove that 😄

FancyKillerPanda commented on src/modules/CurrencyModule.h:
```diff
@@ -4,6 +4,15 @@
 
+
+//Used for the events (Needed in the EventModule.cpp and CurrencyModule.cpp)
+#define highRiskRewardBonus 60
+#define highRiskRewardStealDecrease 8
+
+#define lowRiskRewardPenalty 50
+#define lowRiskRewardLeastAmoutToSteal 60
+#define lowRiskRewardStealIncrease 20
```
These should probably be set-able through the bot itself.

ridhwandll replied:
I think that it will be a lot of admin junk commands, like 5 of them here....when more events added...just imagine...

FancyKillerPanda commented on src/modules/EventModule.cpp:
```diff
@@ -0,0 +1,463 @@
+
+EventModule::EventModule(UmikoBot* client) : Module("event", true), m_client(client)
+{
+	RegisterCommand(Commands::EVENT, "event", [this](Discord::Client& client, const Discord::Message& message, const Discord::Channel& channel)
```
Reuse of `client` could be confusing.

Why do we even need the `UmikoBot* client` parameter? `UmikoBot` was made a singleton so it can just be accessed through `UmikoBot::Instance().whatever`.

FancyKillerPanda commented on src/modules/EventModule.cpp:
```diff
@@ -0,0 +1,463 @@
+			client.createMessage(message.channelId(), "**Wrong Usage of Command!** ");
+			return;
+		}
+		if (config.isEventRunning == false)
```
`if (!config.isEventRunning)`

ridhwandll replied:
Well you guys use that style, didn't know. I am changing that 😄

FancyKillerPanda replied:
FYI it's not really a matter of style, that's just common C++ practice

ridhwandll replied:
Yes perfectly changed that!

FancyKillerPanda commented on src/modules/CurrencyModule.cpp:
```diff
@@ -1198,7 +1202,55 @@ CurrencyModule::CurrencyModule(UmikoBot* client) : Module("currency", true), m_c
+		{
+			QString num = QString::number(lowRiskRewardLeastAmoutToSteal);
+			QString desc = QString(
+				"**You can't steal that amount beacuse LowRiskLowReward event is running!**\n"
+				"*(You have to steal more than or equal to `%2 %1`)*\n"
```
Why is there a minimum amount to steal?

ridhwandll replied:
Because the LowRiskLowReward event issues a penalty if the user succeeds in stealing! The least amount is there to stop losing money by stealing, suppose you steal 10 CC from a person, you succeeded, now in LowRiskLowReward event you will be charged 50 CC (approximately) for succeeding so you will lose 40 CC if you try to steal 10 CC and succeed! To prevent this the least steal amount is introduced in LowRiskLowReward event! 🙂

FancyKillerPanda replied:
Shouldn't the penalty be proportional to the amount stolen?

ridhwandll replied:
Nice idea, I will try to implement that!

ridhwandll replied:
Done!

FancyKillerPanda commented on src/modules/CurrencyModule.cpp:
```diff
@@ -1198,7 +1202,55 @@ CurrencyModule::CurrencyModule(UmikoBot* client) : Module("currency", true), m_c
+		if (roll && eventConfig.eventLowRiskLowRewardRunning)
+		{
+			QString num = QString::number(lowRiskRewardPenalty);
+			victimCurrency.setCurrency(victimCurrency.currency() - amountToSteal);
```
The victim should only lose the amount that the thief steals.

ridhwandll replied:
Yes, @FancyKillerPanda , the victim loses that currency! all good here I think.

ridhwandll replied:
Yeah, at last got it. Thanks for pointing that, that was really a big mistake, sorry for that. It is now fixed. 😄

FancyKillerPanda commented on src/modules/CurrencyModule.cpp:
```diff
@@ -1208,10 +1260,10 @@ CurrencyModule::CurrencyModule(UmikoBot* client) : Module("currency", true), m_c
 		}
-		else
+
+		else if(!roll)
```
`else if (!roll)` after `else if (roll)`?

ridhwandll replied:
Should I use only `else` ?

FancyKillerPanda commented on src/modules/CurrencyModule.cpp:
```diff
@@ -1239,22 +1291,26 @@ CurrencyModule::CurrencyModule(UmikoBot* client) : Module("currency", true), m_c
+				client.createMessage(message.channelId(), "**You cannot set configarations regarding steal while an event is running!** ");
+				return;
+			}
 			if (args.size() == 2)
```
You've already checked for this.

ridhwandll replied:
Didn't get you this time 😋

FancyKillerPanda commented on src/modules/CurrencyModule.cpp:
```diff
@@ -1270,22 +1326,26 @@ CurrencyModule::CurrencyModule(UmikoBot* client) : Module("currency", true), m_c
+				client.createMessage(message.channelId(), "**You cannot set configarations regarding steal while an event is running!** ");
+				return;
+			}
 			if (args.size() == 2)
```
You've already checked for this.

ridhwandll replied:
Sorry I didn't get you, I have checked this once in this command.

FancyKillerPanda commented on src/modules/CurrencyModule.cpp:
```diff
@@ -1301,22 +1361,26 @@ CurrencyModule::CurrencyModule(UmikoBot* client) : Module("currency", true), m_c
+				client.createMessage(message.channelId(), "**You cannot set configarations regarding steal while an event is running!** ");
+				return;
+			}
 			if (args.size() == 2)
```
You've already checked for this.

ridhwandll replied:
Sorry I didn't get you, I have checked this once in this command.

FancyKillerPanda commented on src/modules/CurrencyModule.cpp:
```diff
@@ -1332,32 +1396,35 @@ CurrencyModule::CurrencyModule(UmikoBot* client) : Module("currency", true), m_c
+				client.createMessage(message.channelId(), "**You cannot set configarations regarding steal while an event is running!** ");
+				return;
+			}
 			if (args.size() == 2)
```
You've already checked for this.

ridhwandll replied:
Sorry I didn't get you, I have checked this once in this command.

FancyKillerPanda commented on src/modules/EventModule.h:
```diff
@@ -0,0 +1,33 @@
+private:
+	QMap<snowflake_t, QList<snowflake_t>> m_EventWhitelist;
+	QMap<snowflake_t, EventConfig>serverEventConfig;
+	UmikoBot* m_client;
```
Why do we need this when `UmikoBot` is a singleton?

ridhwandll replied:
Yes perfectly removed!

FancyKillerPanda commented on src/modules/FunModule.cpp:
```diff
@@ -70,31 +70,36 @@ FunModule::FunModule(UmikoBot* client) : Module("funutil", true), m_memeChannel(
 			return;
 		}
 
+		double min = args.at(1).toDouble();
+		double max = args.at(2).toDouble();
+
 		QRegExp re("[+-]?\\d*\\.?\\d+");
 		if (!re.exactMatch(args.at(1)) || !re.exactMatch(args.at(2)))
```
Should probably check if they are doubles before you try converting them.

ridhwandll replied:
Any need for that? If yes, then how? 😄

FancyKillerPanda commented on src/modules/FunModule.cpp:
```diff
@@ -70,31 +70,36 @@ FunModule::FunModule(UmikoBot* client) : Module("funutil", true), m_memeChannel(
 		}
 
+		//Max and min in 32-bit is 2147483647 and -2147483647
+		if (max >= 2147483647 || min >= 2147483647 || max <= -2147483647 || min <= -2147483647)
```
Use the constants for `INT_MAX`, etc.

Also why checking `>=` and `<=` instead of `>` and `<`?
Also the negative ones are incorrect, `-2147483648` should be valid.

ridhwandll replied:
Sure, done the changes!

FancyKillerPanda commented on src/modules/FunModule.cpp:
```diff
@@ -70,31 +70,36 @@ FunModule::FunModule(UmikoBot* client) : Module("funutil", true), m_memeChannel(
+		//Max and min in 32-bit is 2147483647 and -2147483647
+		if (max >= 2147483647 || min >= 2147483647 || max <= -2147483647 || min <= -2147483647)
+		{
+			client.createMessage(message.channelId(), "**You can't roll that number!**");
```
This should be more descriptive, *why* can't they roll that number...

ridhwandll replied:
Will I tell them that it's beyond 32-bit limit? I personally don't like to tell them this 32-bit thingy.

FancyKillerPanda commented on src/modules/EventModule.cpp:
```diff
@@ -0,0 +1,463 @@
+						return;
+					}
+					snowflake_t guild = channel.guildId();
+					snowflake_t chan = channel.id();
```
This is never used.

ridhwandll replied:
The `guild` is used, but the `chan` is not, I am removing the `chan` 🙂

FancyKillerPanda commented on src/modules/EventModule.cpp:
```diff
@@ -0,0 +1,463 @@
+
+				});
+		});
+	RegisterCommand(Commands::EVENT_TAKE_NEW_ACCESS, "take-new-event-access", [this](Discord::Client& client, const Discord::Message& message, const Discord::Channel& channel)
```
Take *new* event access?

ridhwandll replied:
Yes, take event launching and ending access from the people who have the access! (Works like the polls)

FancyKillerPanda commented on src/modules/EventModule.cpp:
```diff
@@ -0,0 +1,463 @@
+					}
+
+					snowflake_t guild = channel.guildId();
+					snowflake_t chan = channel.id();
```
Never used

FancyKillerPanda commented on src/modules/EventModule.cpp:
```diff
@@ -0,0 +1,463 @@
+					}
+					if (time > 24)
+					{
+						UmikoBot::Instance().createMessage(message.channelId(), "**Currently you can't launch events with an expiry time of more than one day**.");
```
Didn't we fix this?

ridhwandll replied:
Nope, @Sid200 didn't open a PR, he fixed that! 😄

FancyKillerPanda commented on src/modules/EventModule.cpp:
```diff
@@ -0,0 +1,463 @@
+					config.eventTimer->setInterval(time * 3600000);
+					config.eventTimer->setSingleShot(true);
+
+					if (args.at(1) == "HighRiskHighReward")
+					{
+						QString num = QString::number(time);
+
+						Discord::Embed embed;
+						embed.setColor(15844367);
+						embed.setTitle("**HighRiskHighReward** event has been launched for `" + num + "` hour(s)!");
+						embed.setDescription("Give command `!event` to see what event is running and its changes!");
+
+						config.isEventRunning = true;
+						config.eventHighRiskHighRewardRunning = true;
+						currencyConfig.stealSuccessChance -= highRiskRewardStealDecrease;
+						UmikoBot::Instance().createMessage(message.channelId(), embed);
+					}
+					if (args.at(1) == "LowRiskLowReward")
```
What if neither of these are true?

ridhwandll replied:
Hmm gotta add that, thanks for pointing that out! 😄

FancyKillerPanda commented on src/modules/EventModule.cpp:
```diff
@@ -0,0 +1,463 @@
+
+								config.isEventRunning = false;
+								config.eventHighRiskHighRewardRunning = false;
+								currencyConfig.stealSuccessChance += highRiskRewardStealDecrease;
+								UmikoBot::Instance().createMessage(chan, embed);
+							}
+							else if (config.eventLowRiskLowRewardRunning)
+							{
+								Discord::Embed embed;
+								embed.setColor(15158332);
+								embed.setTitle("**LowRiskLowReward** event has **ended**!");
+
+								config.isEventRunning = false;
+								config.eventLowRiskLowRewardRunning = false;
+								currencyConfig.stealSuccessChance -= lowRiskRewardStealIncrease;
```
You shouldn't modify the actual steal success chance variable, as that is saved every few minutes (and it represents the normal steal success chance).

ridhwandll replied:
I restore the real steal success chance every time the bot starts (after crash)!

FancyKillerPanda replied:
I still feel like it'd be safer if we just didn't modify those, let's see what @Gaztin says

ridhwandll replied:
Can we make a copy of the real `stealSuccessChance` and store that in a variable and tweak the copied variable?

Gaztin replied:
Yeah don't modify those. What you can do is change how we calculate the success chance when a user actually performs a steal command, and use the event system there 😊

ridhwandll replied:
Okay 😄

FancyKillerPanda commented on src/modules/EventModule.cpp:
```diff
@@ -0,0 +1,463 @@
+					UmikoBot::Instance().createMessage(message.channelId(), "**Wrong Usage of Command!** ");
+					return;
+				}
+				if (config.isEventRunning == false)
```
`if (!config.isEventRunning)`

FancyKillerPanda commented on src/modules/EventModule.cpp:
```diff
@@ -0,0 +1,463 @@
+				auto& config = getEventServerData(guildId);
+				auto& currencyConfig = currencyModule->getServerData(channel.guildId());
+
+				if (config.restored == true)
+				{
+					return;
+				}
+
+				if (config.isEventRunning == false)
+				{
+					config.restored = true;
+					return;
+				}
+				if (config.restored == false)
```
You can simplify a lot of these `if`s.

If you've already checked for one condition, you (usually) don't need to check it again.

ridhwandll replied:
Sure!

ridhwandll committed:
Now selected roles can launch or end events.


Fix some responses in events


Typo fixed in the bribe command!


Some minor fixes and optimizations!

ridhwandll committed:
Some fixes regarding the "!roll" and the event system

ridhwandll committed:
In events, the real stealSuccessChance is no longer tweaked.

ridhwandll committed:
Merge remote-tracking branch 'upstream/master'

ridhwandll committed:
Some fix after the merge.

FancyKillerPanda commented on res/commands.json:
```diff
@@ -270,14 +270,40 @@
 	"FUN_ROLL": {
 		"brief": "Gets a random number.",
-		"usage": "roll <min> <max>",
+		"usage": "roll <min> <max> or !roll <max>",
```
I think these should be on separate lines, sort of like how `!help top` does it.

FancyKillerPanda commented on src/modules/CurrencyModule.cpp:
```diff
@@ -915,20 +917,123 @@ CurrencyModule::CurrencyModule(UmikoBot* client) : Module("currency", true), m_c
+		{
+			double successChance = (config.highRiskRewardStealSuccessChance / 100.0) * qExp(-0.0001 * qPow(amountToSteal, 1.5));
+
+			QString thiefName = UmikoBot::Instance().GetName(channel.guildId(), authorId);
+			QString victimName = UmikoBot::Instance().GetName(channel.guildId(), victimId);
```
You already created these above.

FancyKillerPanda commented on src/modules/CurrencyModule.cpp:
```diff
@@ -915,20 +917,123 @@ CurrencyModule::CurrencyModule(UmikoBot* client) : Module("currency", true), m_c
+		{
+			double successChance = (config.lowRiskRewardStealSuccessChance / 100.0) * qExp(-0.0001 * qPow(amountToSteal, 1.5));
+
+			QString thiefName = UmikoBot::Instance().GetName(channel.guildId(), authorId);
+			QString victimName = UmikoBot::Instance().GetName(channel.guildId(), victimId);
```
You already created these above.

FancyKillerPanda commented on src/modules/CurrencyModule.cpp:
```diff
@@ -915,20 +917,123 @@ CurrencyModule::CurrencyModule(UmikoBot* client) : Module("currency", true), m_c
+		EventModule* eventModule = static_cast<EventModule*>(UmikoBot::Instance().GetModuleByName("event"));
+		auto& eventConfig = eventModule->getEventServerData(channel.guildId());
+
+		if (eventConfig.eventHighRiskHighRewardRunning)
```
(This comment applies to a lot of the steal function, i haven't quoted it all for compactness sake).

I feel like it would be better to calculate the variables once and then have the steal logic execute to avoid code duplication. Instead of copy-pasting the steal code in these `if` statements a bunch of times, we should instead create local variables for the success chance, victim bonus, thief penalty, etc. Then have the steal logic run with those variables, so you don't need to have the steal logic copied many times here.

ridhwandll replied:
Yes, I have done this (kind of in the new PR)

FancyKillerPanda commented on src/modules/CurrencyModule.cpp:
```diff
@@ -915,20 +917,123 @@ CurrencyModule::CurrencyModule(UmikoBot* client) : Module("currency", true), m_c
+				QString output = QString(
+					":man_detective: **Steal success!** :man_detective:\n"
+					"*%1* has discreetly stolen **`%2 %3`** from under *%4's* nose.\n"
+					"HighRiskHighReward event **BONUS**: `%5 %3`\n"
```
You can just append this part of the message to the `outputString` or whatever after the main portion of the steal logic is done.

FancyKillerPanda commented on src/modules/CurrencyModule.cpp:
```diff
@@ -915,20 +917,123 @@ CurrencyModule::CurrencyModule(UmikoBot* client) : Module("currency", true), m_c
+			std::mt19937 prng{ device() };
+			std::discrete_distribution<> distribution{ { 1 - successChance, successChance } };
+			int roll = distribution(prng);
+			qDebug() << "local success chance"<< successChance;
+			qDebug() << "steal success chance"<< config.stealSuccessChance;
```
These shouldn't be here in the final build.

ridhwandll replied:
Yea, removed these!

FancyKillerPanda commented on src/modules/CurrencyModule.cpp:
```diff
@@ -915,20 +917,123 @@ CurrencyModule::CurrencyModule(UmikoBot* client) : Module("currency", true), m_c
+			std::discrete_distribution<> distribution{ { 1 - successChance, successChance } };
+			int roll = distribution(prng);
+
+			qDebug() << "local success chance" << successChance;
+			qDebug() << "steal success chance" << config.stealSuccessChance;
```
Same as above

ridhwandll replied:
Same as above

FancyKillerPanda reviewed (changes requested):
Please make sure you test that everything still works after rebasing.
